pkg/duplex/server_test.go

```diff
@@ -0,0 +1,209 @@
+/*
+Copyright 2025 Chainguard, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package duplex
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	pb "chainguard.dev/go-grpc-kit/pkg/duplex/internal/proto/helloworld"
+	"chainguard.dev/go-grpc-kit/pkg/metrics"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+func GetUnusedPort(t *testing.T) int {
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer listener.Close()
```

Comment:
This potentially could race because you're grabbing a port, to then defer the close in the scope of this func which then frees the port, to then try to re-listen on the port later.

Ideally we'd pass the listener through, but idk how much refactoring you want to do in this PR. 🤷

Reply:
ack'ed

added a comment, i would like to get this in as is, and if it becomes flaky, happy to refactor